### PR TITLE
Get the list of auth methods supported by the server when first rendering the auth form

### LINKED ISSUE #331: Fetch authentication methods from server capabilities
* User picks a server
* Capabilities are fetched
* Known authenticated methods are filtered against the methods that the server supports

### PR DESCRIPTION
**Pull request by magopian**
Fixes #331

### CODE REVIEW

**Comment by magopian** [comment]
This isn't ready yet, it also needs the capabilities to be parsed on each server change (the current `onChange` implementation is just some placeholder for now).

File: src/components/AuthForm.js [hunk 1]
```diff
@@ -363,18 +367,67 @@ export default class AuthForm extends PureComponent<
...
+    const knownAuthImplementations = [
+      "basicauth",
+      "account",
+      "fxa",
+      "ldap",
+      "portier",
+      "anonymous",
+    ];
+    const server = (history && history[0]) || defaultKintoServer;
+    const defaultAuth = "anonymous";
```
Comment (leplatrem) [inline src/components/AuthForm.js]: nit: this could be a global const

File: src/components/AuthForm.js [hunk 2]
```diff
@@ -363,18 +367,67 @@ export default class AuthForm extends PureComponent<
...
+      // Server changed, request its capabilities and check what auth methods
+      // it supports.
+      setupClient({ authType: "anonymous", server: server });
+      const client = getClient();
+      client
+        .fetchServerInfo()
+        .then(info => {
+          console.log("fetchServerInfo", info);
+        })
+        .catch(() => console.log("error"));
```
Comment (leplatrem) [inline src/components/AuthForm.js]: nit: use `await` (and possibly reuse code with `componentDidMount`

Comment (magopian) [reply src/components/AuthForm.js]: yeah, as explained above, this code is only a placeholder (a previous version of the code i didn't mean to commit...) I'll finalize this tomorrow ;)

**Comment by magopian** [comment]
Regarding the use of redux/sagas, i'm not sure... it seems that most of the state is handled in the component when it is **local** (like the auth methods in the `AuthForm` I guess?).

In a project using redux/sagas for everything I would have done it, in this specific case, I'm not sure... What do you think? Also maybe @glasserc has some opinion on that?


Oh, also this PR needs tests.

**Comment by glasserc** [comment]
I'm probably the wrong person to ask about these things, because I don't really understand why components in React have local state. Up until you said anything, I thought everything in kinto-admin *did* use Redux/Sagas. Now I'm not sure -- I guess anything with a `setState()` call is skirting it. I feel like hewing closer to Redux/Sagas is probably better than hewing further away from it, but I guess Redux/Saga probably also imposes a certain amount of boilerplate and that is unpleasant? I guess I feel like maybe avoiding Redux/Sagas in some cases is OK, but probably talking to servers should be done using Redux/Sagas.

**Comment by leplatrem** [comment]
> . it seems that most of the state is handled in the component when it is local (like the auth methods in the AuthForm I guess?).

Yes, if I remember well what @n1k0 told me, we use the local state only for minor things like toggles, list expansion etc. 

> In a project using redux/sagas for everything I would have done it, in this specific case, I'm not sure... 

I would say that kinto-admin uses sagas for everything. If you spotted a place that makes this statement partly incorrect, then please open an issue we can track it as tech-debt :)

> Redux/Saga probably also imposes a certain amount of boilerplate and that is unpleasant?

Oh yeah indeed. It also requires to «name» many things like actions, global state attributes etc. But in the end, if it's consistent with the rest, it make sense

>  talking to servers should be done using Redux/Sagas

Yes, agree. And also for important side effects like calling `setClient()`

**Comment by magopian** [comment]
I moved the code to a saga, which will be easier to test (I'll push a commit soon).
The current implementation still has issues (need to remove all reference to `authMethods` in the `AuthForm` local state for example, and from the `SettingsState`, and fix some undefined issues, I love JS).

**Comment by magopian** [comment]
Ok, I think this is now ready to be merged. There's one caveat left: whenever you make a change in the server input, it'll try to fetch the server info.

This is not optimal, but mitigated by the fact that the kinto-http.js library won't do a fetch unless the url "looks good" (eg: you have the url schema, the version etc...). Also, if you just select the server from the history, this won't be an issue.

We might want to fix that though in the future, and at the same time fix the fact that whenever there's a change event, the cursor is placed at the end of the input (if you try to add a few letters in the middle of the url, you'll have a hard time because after each keypress the cursor is moved to the end of the string).

@leplatrem @glasserc what do you think, is that good enough to be merged now?

**Comment by leplatrem** [comment]
Quick reply regarding the onchange event, you can just debounce the event, and fetch the server info only when no character was typed for like 500ms for example

File: src/components/AuthForm.js [hunk 3]
```diff
@@ -6,6 +6,9 @@ import React, { PureComponent } from "react";
 import BaseForm from "./BaseForm";
 import { omit } from "../utils";
 
+const defaultKintoServer = "https://kinto.dev.mozaws.net/v1/";
+const defaultAuth = "anonymous";
```
Comment (leplatrem) [inline src/components/AuthForm.js]: I think this constant could be named «anonymous auth» explicitly

File: src/components/AuthForm.js [hunk 4]
```diff
@@ -363,18 +366,49 @@ export default class AuthForm extends PureComponent<
...
+    const { session: { serverInfo: { capabilities } } } = this.props;
+    // Check which of our known auth implementations are supported by the server.
+    const knownAuthMethods = [
+      "basicauth",
+      "account",
+      "fxa",
+      "ldap",
+      "portier",
+      "anonymous",
+    ];
```
Comment (leplatrem) [inline src/components/AuthForm.js]: nitpick: could be const on top of source file

File: src/components/AuthForm.js [hunk 5]
```diff
@@ -363,18 +366,49 @@ export default class AuthForm extends PureComponent<
...
+    let supportedAuthMethods = [defaultAuth];
+    knownAuthMethods.forEach(authMethod => {
+      if (authMethod in capabilities) {
+        supportedAuthMethods.push(authMethod);
+      }
+    });
```
Comment (leplatrem) [inline src/components/AuthForm.js]: nitpick: `knownAuthMethods.filter(a => a in capabilities)` ?

File: src/components/AuthForm.js [hunk 6]
```diff
@@ -363,18 +366,49 @@ export default class AuthForm extends PureComponent<
...
+  getSupportedAuthMethods = () => {
+    const { session: { serverInfo: { capabilities } } } = this.props;
+    // Check which of our known auth implementations are supported by the server.
+    const knownAuthMethods = [
+      "basicauth",
+      "account",
+      "fxa",
+      "ldap",
+      "portier",
+      "anonymous",
```
Comment (leplatrem) [inline src/components/AuthForm.js]: Note: There is not `anonymous` capability (can be removed from list IMO)

File: src/components/AuthForm.js [hunk 7]
```diff
@@ -6,6 +6,9 @@ import React, { PureComponent } from "react";
 import BaseForm from "./BaseForm";
 import { omit } from "../utils";
 
+const defaultKintoServer = "https://kinto.dev.mozaws.net/v1/";
```
Comment (leplatrem) [inline src/components/AuthForm.js]: nitpick: uppercase variables names?

File: src/components/AuthForm.js [hunk 8]
```diff
@@ -363,18 +366,49 @@ export default class AuthForm extends PureComponent<
...
+    return supportedAuthMethods;
+  };
+
   onChange = ({ formData }: { formData: Object }) => {
-    const { authType } = formData;
+    const { authType, server } = formData;
+    if (this.state.formData.server !== server) {
+      // Server changed, request its capabilities to check what auth methods it
+      // supports.
+      const anonymousAuth = { authType: "anonymous", server: server };
```
Comment (leplatrem) [inline src/components/AuthForm.js]: could use const anonymous type maybe

File: src/sagas/session.js [hunk 9]
```diff
@@ -15,25 +15,46 @@ import { saveSession, clearSession } from "../store/localStore";
...
     const serverInfo = yield call([client, client.fetchServerInfo]);
+    // Notify they're received
+    yield put(actions.serverInfoSuccess(serverInfo));
+
+    yield put(notificationActions.clearNotifications({ force: true }));
+  } catch (error) {
+    // Reset the server info that we might have added previously to the state.
+    yield put(actions.serverInfoSuccess(DEFAULT_SERVERINFO));
+    yield put(
+      notificationActions.notifyError("Couldn't retrieve server info", error)
```
Comment (leplatrem) [inline src/sagas/session.js]: `Could not reach server` maybe?

File: test/sagas/session_test.js [hunk 10]
```diff
@@ -37,21 +39,77 @@ describe("session sagas", () => {
...
+        expect(getServerInfo.next().value).eql(
+          put(notificationsActions.clearNotifications({ force: true }))
+        );
+      });
     });
 
-    it("should configure the client", () => {
-      expect(getClient().remote).eql(authData.server);
+    describe("Failure", () => {
+      it("should reset the server info", () => {
```
Comment (leplatrem) [inline test/sagas/session_test.js]: just out of curiosity, why should we reset server info?

Comment (magopian) [reply test/sagas/session_test.js]: If you enter a bogus server in the input box, you don't want to still have the capabilities from the previous "working" server. Does that answer your question? Did I understand your question correctly?

File: src/components/AuthForm.js (hunk 8 above)
Comment (magopian) [reply src/components/AuthForm.js]: What do you mean?

File: src/components/AuthForm.js (hunk 5 above)
Comment (magopian) [reply src/components/AuthForm.js]: And then not forget to add the "anonymous" method which isn't in the capabilities as you said. Looks like a good idea, thanks ;)

File: test/sagas/session_test.js (hunk 10 above)
Comment (leplatrem) [reply test/sagas/session_test.js]: yep! got it, wasn't obvious from the reviewer standpoint :)

Comment (magopian) [reply test/sagas/session_test.js]: Good point, I'll add a comment in the code ;) Thanks for bringing it up!

File: src/components/AuthForm.js (hunk 8 above)
Comment (leplatrem) [reply src/components/AuthForm.js]: Something like `const ANONYMOUS_AUTH = "anonymous"` ... but that's kind of redundant with the other comments regarding this anonymous string

Comment (magopian) [reply src/components/AuthForm.js]: Ok, I added a ANONYMOUS_AUTH_DATA utility.

**Comment by magopian** [comment]
@leplatrem I believe I've addressed your comments, what do you think?

**Pull request merged**